BEST_PRACTICES.md

```diff
@@ -0,0 +1,84 @@
+has been created to simplify this, but you'll still want to be careful
+with that functionality due to the open security it implies.
+* `input` shouldn’t be done within a `node` block - that eats up two
+executors waiting on the `input` feedback, both the flyweight executor
+that’s used for the `input` itself and the real executor used by the
+`node` block, which won’t free up until after the `input` itself has completed.
+
+# Parallelism
+* Within `parallel` blocks, use `node` blocks to make sure you farm out to
+real nodes for your parallelized work.
+* Nested `parallel` blocks can lead to swamping your available
+executors, as each execution of the first `parallel` block calls
+multiple executions of the second `parallel` block, and so on. In
+general, think carefully about your parallelism and your available executors
+when using `parallel`.
+* The [Parallel Test Executor
+plugin](https://github.com/jenkinsci/parallel-test-executor-plugin) is
+awesome and can be immensely helpful both for distributing your test
+execution and for throttling your parallelism, since you define how
+many "buckets" your tests get divided into.
+* Don’t put `stage`s in `parallel` blocks - that just goes *weird*,
+breaking a lot of logic in the Stage View and elsewhere. Save yourself
+the pain - don't do it!
+
+# `Jenkinsfile`s and Multibranch
+* Use: `checkout scm` to automatically checkout current revision of branch
+* Use: `$env.BRANCH_NAME` variable if you have logical difference in
+your flow between branches, i.e. to distinguish different behavior for
+production-ready branches versus sandbox or pull request branches.
+* For `Jenkinsfile`s, make sure to put `#!groovy` at the top of the
+file so that IDEs, GitHub diffs, etc properly detect the language and
+do syntax highlighting for you.
+
+# Groovy gotchas
+* Don’t have the Groovy interpreter making blocking i/o calls, i.e.,
+`HTTPClient` and the like - these can cause real problems with
+resumability, and also require a lot of explicit whitelisting of
+methods in the Script Security plugin, which is not ideal.
+* Beware `for (Foo f: foos)` loops and Groovy closure-style operators
+like `.each` and the like. They will not work right in normal Workflow
+script contexts where Workflow steps are involved directly.
+* If you need to do that kind of thing to make your scripting make
+sense, do it in methods annotated with `@NonCPS`. These methods will
+not be CPS-transformed, so can do some things that just can't be done
+in the serializable/resumable CPS context.
+* Don’t use the Groovy scripting in place of shell scripting - work
+coming for the ability to run a Groovy step on the node as with the
+normal Groovy plugin build step, but until then, shell out, even if
+it’s just to do `sh 'groovy foo.groovy'`.
+* If you really need `Map`s in your normal Workflow scripts, consider
+using arrays of `[key,value]` instead - arrays have serializable
```

Comment:
The reason for the `[[key,value], [key2,value2], ... ]` arrays isn't to do with arrays having serializable iterators (they don't, you'll have issues with the `for (Foo f: foos)` as noted above) -- it's because this construct allows for C-style for loops, fetching element by element using an index.